src/instructlab/eval/mt_bench_branch_generator.py

```diff
@@ -22,10 +23,20 @@
 def get_file_paths(directory):
     logger.debug(locals())
     file_paths = []
-    for root, _, files in os.walk(directory):
-        for file in files:
-            if file.split("/")[-1] == "qna.yaml":
-                file_paths.append(os.path.join(root, file))
+    root_paths = [
+        entry
+        for entry in Path(directory).iterdir()
+        if entry.is_dir()
+        if not entry.name.startswith(".")
```

Comment:
extracting the same attribute 4 times... CPU is cheap! :D

Reply:
Now I am curious what would be faster?  Ignoring that the compiler/interpreter will likely rewrite it anyway.  I would bet declaring a variable or an array are both slower.

Reply:
```
$ cat /tmp/test.py
from pathlib import Path
import timeit

paths = [
    Path('test')
    for i in range(10000)
]

FORBIDDEN = {"1", "2", "3"}


def test1():
    return [
        path
        for path in paths
        if not path.name.startswith('.')
        if path.name != "1"
        if path.name != "2"
        if path.name != "3"
    ]


def test2():
    return [
        path
        for path in paths
        if not path.name.startswith('.')
        if path.name not in FORBIDDEN
    ]


print(timeit.timeit(test1, number=1000))
print(timeit.timeit(test2, number=1000))
$ python /tmp/test.py
6.865049810003256
4.405094917005044
```

CPython is not very smart.

Reply:
In recent pythons, we can even squash the last duplicate attribute extraction as in `test4` below:

```
$ cat /tmp/test.py
from pathlib import Path
import timeit

paths = [
    Path('test')
    for i in range(10000)
]

FORBIDDEN = {"1", "2", "3"}
FORBIDDEN_LIST = ["1", "2", "3"]


def test1():
    return [
        path
        for path in paths
        if not path.name.startswith('.')
        if path.name != "1"
        if path.name != "2"
        if path.name != "3"
    ]


def test2():
    return [
        path
        for path in paths
        if not path.name.startswith('.')
        if path.name not in FORBIDDEN
    ]


def test3():
    return [
        path
        for path in paths
        if not path.name.startswith('.')
        if path.name not in FORBIDDEN_LIST
    ]


def test4():
    return [
        path
        for path in paths
        if not (name := path.name.startswith('.'))
        if name not in FORBIDDEN
    ]


print(timeit.timeit(test1, number=1000))
print(timeit.timeit(test2, number=1000))
print(timeit.timeit(test3, number=1000))
print(timeit.timeit(test4, number=1000))
$ python /tmp/test.py
7.211299414993846
5.215823543010629
4.967952872990281
2.9443017129960936
```

-60% CPU, not too shabby. ;)

Reply:
This made my day!  Thanks for sharing.  I am surprised at the set being slower than the list (or even a statistical tie).  Normally a set has a O(1) lookup and the list would be O(n) with the set being a little more expensive to create.

Not sure what makes python attribute lookup so slow either.

Pedantic notes: I don't know if this conclusively proves which is faster still in the real life example.  The actual number of dirs is 7, only 2 of the values make it through all the ifs, and we aren't taking the set/list creation into account in this test.  Given how slow the attr lookup seems to be though, I am guessing 4) would edge out 1) with the real data/code.

Obviously the performance doesn't really matter in this case.  I would personally still pick 3 or 4 from a readability point of view.

Thanks again for putting this together!  I think it's a great practice to really understand how things are working.

Reply:
@danmcp I am happy it was delightful to you too. :D If I may, I will indulge into this meaningless exercise then a bit more. :)

---

> I am surprised at the set being slower than the list (or even a statistical tie).

I think the culprit is the size of the structure. Sets are backed by hash tables, which, in a degenerate case of a short iterable, may have a similar or even higher overhead. (One has to calculate the hash function, then find the relevant bucket and perhaps even walk the bucket list.)

I [bumped](https://gist.github.com/booxter/0b16403775a131b6a88f4db1c3238df8) the size of the set of forbidden values from 3 to 100, and I get more "reasonable" numbers in terms of list vs. set(s) behavior (I also added `frozenset` to the mix - it seems to do slightly better than a regular set but not by much.)

```
if-if-if 150.47323519000201
in-set 4.468321249005385
in-list 13.755651542975102
in-frozenset 4.645498502010014
in-set-one-attr-fetch 3.846038645016961
in-list-one-attr-fetch 13.177363599010278
in-frozenset-one-attr-fetch 3.4668387069832534
```

> Not sure what makes python attribute lookup so slow either.

I think this may have something to do with the fact that `Path.name` is not a simple (or even `__slots__`) attribute but a [property](https://github.com/python/cpython/blob/fb864c76cd5e450e789a7b4095832e118cc49a39/Lib/pathlib/_local.py#L347) (meaning, a function is called on each access to the attribute). The property implementation is not huge, but it makes another `__getattr__` and a `__getitem__` calls. Perhaps a "true" attribute would fare differently. (But I have to stop falling into this rabbit hole, now!)